torch/backends/_nnapi/prepare.py

```diff
@@ -0,0 +1,167 @@
+        comp.run(fixed_args, outs)
+        assert len(outs) == len(self.out_mem_fmts)
+        for idx in range(len(self.out_templates)):
+            fmt = self.out_mem_fmts[idx]
+            if fmt == 0:
+                pass
+            elif fmt == 1:
+                outs[idx] = outs[idx].permute(0,3,1,2)
+            else:
+                raise Exception("Invalid mem_fmt")
+        return outs
+    
+
+class NnapiInitWrapper(torch.nn.Module):
+    """Wrapper module to ensure NNAPI init is called."""
+    def __init__(self, nnapi_module):
+        super().__init__()
+        self.nnapi_module = nnapi_module
+        
+    def forward(self, args: List[torch.Tensor]) -> List[torch.Tensor]:
+        return self.nnapi_module(args)
+    
+    @torch.jit.export
+    def __getstate__(self):
+        return self.nnapi_module
+        
+    @torch.jit.export
+    def __setstate__(self, nnapi_module):
+        self.training = False
+        self.nnapi_module = nnapi_module
+        self.nnapi_module.init()
+
+
+class ListWrapper(torch.nn.Module):
+    """NNAPI list-ifying wrapper.
+
+    NNAPI always expects a list of inputs.  This module provides a
+    single-tensor input interface for models that want it.
+    """
+    def __init__(self, mod):
+        super().__init__()
+        self.mod = mod
+        
+    def forward(self, t: torch.Tensor) -> List[torch.Tensor]:
+        return self.mod([t])
+        
+class DelistWrapper(torch.nn.Module):
+    """NNAPI de-list-ifying wrapper.
+
+    NNAPI always provides a list of outputs.  This module provides a
+    single-tensor output interface for models that want it.
+    """
+    def __init__(self, mod):
+        super().__init__()
+        self.mod = mod
+        
+    def forward(self, ts: List[torch.Tensor]) -> torch.Tensor:
+        outs = self.mod(ts)
+        assert len(outs) == 1
+        return outs[0]
+
+class ListDelistWrapper(torch.nn.Module):
+    """NNAPI list-ifying and de-list-ifying wrapper.
+
+    NNAPI always expects a list of inputs and provides a list of outputs.
+    This module provides a single-tensor input/output interface
+    for models that want it.
+    """
+    def __init__(self, mod):
+        super().__init__()
+        self.mod = mod
+        
+    def forward(self, t: torch.Tensor) -> torch.Tensor:
+        outs = self.mod([t])
+        assert len(outs) == 1
+        return outs[0]
+
+
+def convert_model_to_nnapi(model, inputs):
+    model = torch.jit.freeze(model)
+
+    outputs = model(inputs)
+
+    if isinstance(inputs, torch.Tensor):
+        nnapi_inputs = [inputs]
+        list_inputs = True
+    else:
+        list_inputs = False
+
+    if isinstance(outputs, torch.Tensor):
+        outputs = [outputs]
+        delist_outputs = True
+    else:
+        delist_outptus = False
+
+    ser_model, used_weights, inp_mem_fmts, out_mem_fmts = serialize_model(model, nnapi_inputs)
+    ser_model_tensor = torch.tensor(list(ser_model), dtype=torch.uint8)
```

Comment:
I don't understand this bit. The serialized model is a character stream (string) right? What is the purposes of converting it into a Tensor?

Reply:
I trust PyTorch to do a better job serializing int8 tensors than strings.